library/mps/reader.c

```diff
@@ -0,0 +1,626 @@
+{
+    unsigned char *frag, *acc;
+    size_t aa, end, fo, shift;
+    TRACE_INIT( "reader_commit" );
+
+    /* Check that the reader is in consuming mode. */
+    frag = rd->frag;
+    if( frag == NULL )
+        RETURN( MBEDTLS_ERR_READER_UNEXPECTED_OPERATION );
+
+    acc = rd->acc;
+    end = rd->end;
+
+    if( acc == NULL )
+    {
+        TRACE( trace_comment, "No accumulator, just shift end" );
+        rd->commit = end;
+        RETURN( 0 );
+    }
+
+    fo = rd->acc_share.frag_offset;
+    if( end >= fo )
+    {
+        TRACE( trace_comment, "Started to serve fragment, get rid of accumulator" );
+        shift = fo;
+        aa = 0;
+    }
+    else
+    {
+        TRACE( trace_comment, "Still serving from accumulator" );
+        aa = rd->acc_avail;
+        shift = end;
+        memmove( acc, acc + shift, aa - shift );
+        aa -= shift;
+    }
+
+    end -= shift;
+    fo -= shift;
+
+    rd->acc_share.frag_offset = fo;
+    rd->acc_avail = aa;
+    rd->commit = end;
+    rd->end = end;
+
+    TRACE( trace_comment, "Final state: (end=commit,fo,avail) = (%u,%u,%u)",
+           (unsigned) end, (unsigned) fo, (unsigned) aa );
+    RETURN( 0 );
+}
+
+int mbedtls_reader_reclaim( mbedtls_reader *rd, size_t *paused )
+{
+    unsigned char *frag, *acc;
+    size_t pause, commit;
+    size_t al, fo, fl;
+    TRACE_INIT( "reader_reclaim" );
+
+    if( paused != NULL )
+        *paused = 0;
+
+    frag = rd->frag;
+    if( frag == NULL )
+        RETURN( MBEDTLS_ERR_READER_UNEXPECTED_OPERATION );
+
+    acc    = rd->acc;
+    pause  = rd->pause;
+    commit = rd->commit;
+    fl     = rd->frag_len;
+
+    if( acc == NULL )
+        fo = 0;
+    else
+        fo = rd->acc_share.frag_offset;
+
+    if( pause == 0 )
+    {
+        TRACE( trace_comment, "No unsatisfied read-request has been logged." );
+        /* Check if there's data left to be consumed. */
+        if( commit < fo || commit - fo < fl )
+        {
+            TRACE( trace_comment, "There is data left to be consumed." );
+            rd->end = commit;
+            RETURN( MBEDTLS_ERR_READER_DATA_LEFT );
+        }
+        TRACE( trace_comment, "The fragment has been completely processed and committed." );
+    }
+    else
+    {
+        size_t frag_backup_offset;
+        size_t frag_backup_len;
+        TRACE( trace_comment, "There has been an unsatisfied read-request with %u bytes overhead.",
+               (unsigned) pause );
+
+        if( acc == NULL )
+        {
+            TRACE( trace_comment, "No accumulator present" );
+            RETURN( MBEDTLS_ERR_READER_NEED_ACCUMULATOR );
+        }
+        al = rd->acc_len;
+
+        /* Check if the upper layer has already fetched
+         * and committed the contents of the accumulator. */
+        if( commit < fo )
+        {
+            /* No, accumulator is still being processed. */
+            TRACE( trace_comment, "Still processing data from the accumulator" );
+            if( fo + fl < fo              ||
```

Comment:
Isn't this condition the same as "fl < 0", and therefore, always true? Or do you check for a border case when fo+fl is close to the maximum value of size_t?

Reply:
This checks for unsigned integer overflow.

Reply:
Could you please name it accordingly in a separate variable? It helps a lot and saves time for future investigators.

Reply:
@AndrzejKurek What about adding a comment instead? I think it hardens readability to use separate variables for all these temporaries computed solely for the purpose of overflow checking.

Reply:
If I just may give my subjective opinion - the best code is the one that documents itself.
If you'd add a variable named `isOverflown=false`, move the check from conditions to `isOverflown |= <operation>` and just leave an `if(isOverflown || something)` in the conditions - you won't need any comments. But of course - everything is up to you, if you feel that comments will be better, add a comment.

Reply:
Done.